### ISSUE TITLE
[RFC]: OpenAI /v1/live/sessions fullduplex handler

### ISSUE THREAD

**Issue by vraiti** [vllm-project/vllm-omni#8085]
# Motivation.

Support OpenAI-compliant `/v1/live/sessions` as a unified framework for serving both native fullduplex and turn-based voice models via a fullduplex WS interface in vLLM-Omni.

The architecture proposed in this RFC achieves this in a manner completely transparent to the vLLM-Omni Orchestrator and vLLM engine by reusing existing vLLM-Omni and vLLM machinery for engine-side operations and request state management. This not only enables these models for a new and powerful OpenAI API endpoint, but also does so with a stack constrained to the API server handler itself and a handful of `/v1/live/sessions`-specific functions per supported model.

This architecture intentionally avoids reusing components from the current OmniDuplexSessionHandler stack (with the exception of `SileroStreamingVAD`). This is because, in addition to extending the vLLM-Omni API surface and expanding the list of supported fullduplex models, this feature also serves as a PoC of an alternative, simplified fullduplex architecture (associated PR is expected to be <5k LOC, as compared to the OmniDuplexSessionHandler stack's ~25k LOC, including model-specific adapters) with 0 impact on engine and orchestrator code and minimal maintenance burden per supported model, without sacrificing performance or any fullduplex feature support. 

# Proposed Change.

### 2.1 Goals

Support the following client and server events on the `/v1/live/sessions` WebSocket API.

Client events accepted:
- `session.start`
- `session.update`
- `session.input_audio.append`
- `session.input_audio.mute` / `session.input_audio.unmute`
- `session.instructions.append`
- `response.item.create`
- `response.create`
- `session.close`

Server events emitted:
- `session.started`, `session.updated`
- `session.input_audio.muted`, `session.input_audio.unmuted`, `session.instructions.appended`
- `session.output_audio.delta`
- `session.output_transcript.delta`
- `session.input_transcript.delta`
- `session.delegation.created`, `response.event`
- `session.usage.updated`
- `session.closed`
- `error`
- `info`

Support the following models in the initial implementation of `/v1/live/sessions`:
- Qwen3-Omni: external VAD
- Nemotron VoiceChat: external VAD
- MiniCPM-o 4.5: native VAD
- PersonaPlex: native VAD (with `multi_modal_data_bypass: True`)

#### OpenAI API Features
These are events and session configs that, while officially part of the `/v1/live/sessions` API, are considered out-of-scope and will be rejected or ignored by the API server. A reason is given for each.

##### GPT-Live-1-dependent features
Requires a model trained on gpt-live-1's backend-to-speech-model channel, which vLLM-Omni does not support
- Client `session.thinking.append`
- Client `session.commentary.append`
- Session config `session.delegation.client`

##### Persistent and multi-session management
Very doable given that session state is already in the API server, just excessive for an initial implementation
- Session config `session.store`
- `/v1/live/sessions/{session_id}/fork`
- `/v1/live/sessions/{session_id}/content`
- `/v1/live/sessions/{session_id}/attach`

##### Extra transports
Also excessive for an initial implementation
- WebRTC: `POST /v1/live/sessions` with SDP offer
- SIP: `transport.*`

#### vLLM-Omni fullduplex Extensions
These are features supported by the OmniDuplexSessionHandler stack that have no OpenAI /v1/live/sessions counterpart. A separate RFC should propose spec-respecting extensions to the OpenAI API to support these
- VAD hints: `force_listen` and `is_speech`
- Image and video input

### 3.1 Architecture

```text
+-----------------------------------------------------------------------------+
|                               API server                                    |
|  WS /v1/live/sessions                                                       |
|  +-----------------------------------------------------------------------+  |
|  | LiveSession handler                                                   |  |
|  |  - client event dispatch / server event emission                      |  |
|  |  - session config, unsupported_features gate                          |  |
|  |  - conversation history (API-layer state)                             |  |
|  |                                                                       |  |
|  |  audio in --> [vad: external] VAD client ---------- end of speech --+ |  |
|  |           --> [vad: native]   audio_buffer_ms buffer ---------------+ |  |
|  |                                                                     | |  |
|  |                                                    generation turn  | |  |
|  |                                                                     v |  |
|  |  +-------------------------------+     +----------------------------+ |  |
|  |  | LiveSessionProcessor (per     |<--->| resumable request driver   | |  |
|  |  | model; generic ChatML for     |     | append / cancel+resubmit   | |  |
|  |  | external VAD models)          |     +-------------+--------------+ |  |
|  |  +-------------------------------+                   |                |  |
|  |                                                      |                |  |
|  |  audio out <-- output pacer (1x) <-- multimodal_outputs / text        |  |
|  |  transcripts <-- ASR client (input audio, output audio timestamps)    |  |
|  +------------------------------------------------------|----------------+  |
+---------------------------------------------------------|-------------------+
                                                          v
                          AsyncOmni.generate(StreamingInput generator, resumable)
                                                          |
                                                          v
                +-------------------------------------------------------------+
                | Orchestrator (unmodified)                                   |
                |                                                             |
                |   Stage 0 (thinker) ---------> Stage 1 ---------> Stage 2   |
                |     ^                                                       |
                |     | vLLM LogitsProcessor installed in the stage engine    |
                |     | from PipelineConfig.live_session_config               |
                |     |      .logits_processor[stage_id]                      |
                |     | (MiniCPM-o: enforces the duplex chat format)          |
                +-------------------------------------------------------------+

External services (recipes/, separate processes)

   VAD client (ws://)                       ASR client (http://)
        |                                        |
        v                                        v
+-------------------------------+   +-----------------------------------+
| SileroVAD service             |   | Whisper service                   |
| FastAPI wrapper around        |   | separate vLLM process,            |
| SileroStreamingVAD            |   | openai/whisper-* checkpoint       |
+-------------------------------+   +-----------------------------------+
```

#### 3.1.1 Touch Points

**Entrypoints**
- `vllm_omni/entrypoints/openai/api_server.py`: `/v1/live/sessions` WebSocket route; the resolved `PipelineConfig` exposed to the route
- New: LiveSession handler package under `vllm_omni/entrypoints/openai/` (handler, VAD client, ASR client, output pacer, resumable request driver, `LiveSessionProcessor` base and generic ChatML implementation)

**Recipes**
- New: `recipes/` entries for the SileroVAD service (FastAPI wrapper around `vllm_omni/engine/duplex/vad.py:SileroStreamingVAD`) and the Whisper service (separate vLLM process)

**Configs**
- `vllm_omni/config/stage_config.py`: `live_session_config` on `PipelineConfig`; `live_session_config` on the deploy config and its top-level allowlist; folding `logits_processor[stage_id]` into the stage's existing `logits_processors` engine arg

**Model Executors**
- `vllm_omni/model_executor/models/{minicpmo_4_5,qwen3_omni,personaplex,nemotron_voicechat}/pipeline.py`: `live_session_config` per model
- `vllm_omni/model_executor/models/minicpmo_4_5/`: MiniCPM-o `LiveSessionProcessor` and `LogitsProcessor` (the latter runs in the engine worker, so it lives with the model code rather than under entrypoints)
- `vllm_omni/model_executor/models/personaplex/`: PersonaPlex `LiveSessionProcessor`; multimodal processor registration (none exists today), multimodal embedding hook, `model_intermediate_buffer` read/write, `multimodal_outputs` agent codes (see `multi_modal_data_bypass`)

### 3.2 API & Interface Changes

The supported `/v1/live/sessions` surface is the set of client and server events listed under Goals, served on the `/v1/live/sessions` WebSocket.

Each model supporting `/v1/live/sessions` has a `live_session_config` PipelineConfig attribute:
```python
live_session_config:
    vad: str = "external"|"native"
    multi_modal_data_bypass: Bool                # only vad: native
    live_session_processor: LiveSessionProcessor
    audio_buffer_ms: int                        # only vad: native
    logits_processor: map[int]vllm.LogitsProcessor
    unsupported_features:
        events: list[str]                        
        session_configs: list[str]               
```

The semantics of each `live_session_config` key are given in [Key Technical Decisions](#33-key-technical-decisions).

And an optional `deploy_config` object for `/v1/live/sessions`:
```yaml
live_session_config:
    session_lifetime_s: int = 1800
    external_vad:                               # only vad: external
        output_audio_delta_size_ms: int = 500       
        vad_service_url: str = ws://localhost:15151
    asr_service_url: str = http://localhost:15152
```

`session_lifetime_s` gives a hard server-side cutoff for the duration of `/v1/live/sessions` WebSocket sessions.

See "Barge-in" for motivation for `external_vad.output_audio_delta_size_ms`.

`external_vad.vad_service_url` gives the URL of the SileroVAD service that drives external VAD sessions.

`asr_service_url` gives the URL of the ASR service that computes input audio transcripts and output audio transcript timestamps.

#### VAD types
This handler defines two classes of model: external VAD (Qwen3-Omni, Nemotron VoiceChat) and native VAD (MiniCPM-o, PersonaPlex), which is specified by the pipeline's `live_session_config.vad` value:
- `vad: "external"`: Models that are natively turn-based (Qwen3-Omni, Nemotron VoiceChat). Stream client audio into a SileroVAD instance that drives turn detection
- `vad: "native"`: Models that are natively fullduplex (MiniCPM-o, PersonaPlex). Buffer `audio_buffer_ms` ms of audio and feed directly to the model continuously

#### External Predictive Models
This handler requires both SileroVAD and Whisper as API server helpers. These are predictive machine learning models with substantial memory consumption, so running them in the handler would be prohibitively expensive, and vLLM-Omni has no mechanism otherwise for these kinds of "supporting" models. So, we require them to be run externally and served at a network-accessible address (ideally localhost for minimal latency)

Recipes for both will be provided in `recipes/`. Whisper's recipe will just spawn a separate vLLM process serving an `openai/whisper-*` checkpoint, and SileroVAD's recipe will spawn a thin `FastAPI` wrapper around the `vllm_omni/engine/duplex/vad.py:SileroStreamingVAD` class.

The handler confirms both services are reachable and healthy on each `session.start` and logs an error if at any point either service fails.

#### Generation turns and resumable requests
We will use vLLM resumable requests to stream user input to the model. One resumable request per session, appended to once per generation turn, where a generation turn is the point at which input is submitted to AsyncOmni:
- `vad: external`: VAD end of speech or a `response.create` continuation.
- `vad: native`: every `audio_buffer_ms` of input

For state management purposes, the API server still maintains the conversation history in memory as well. When that history must be mutated, the vLLM request is cancelled, the mutation operation is forwarded to the configured `live_session_processor`, and the mutated history is resubmitted to AsyncOmni as a new request. This can happen upon `session.update` of delegation settings (which can change the Responses backend's `instructions` or `tools`), left-trimming when `max_model_len` is exhausted, or external VAD barge-in truncation.

#### ASR Audio Transcription and Timestamping
We require audio transcription with timestamps for `session.output_transcript.delta{start_ms,end_ms}` event fields, `session.input_transcript.delta` event support, and external VAD barge-in truncation.

Whisper will be used to transcribe audio streams. For model output this transcription will be tokenized using the model input processor and substring-matched (to tolerate lossiness of the LLM -> TTS -> ASR -> Tokenizer chain) to the model's output token stream

#### Barge-in and output pacing
To enable barge-in for external VAD models, `session.output_audio.delta` is released at 1x against the actual audio duration in chunks of size `output_audio_delta_size_ms`, regardless of how quickly the model generates (which is often much faster than real-time playback). If the server VAD detects speech in the middle of streaming output, it can submit a truncation to `live_session_processor` to "interrupt" the model response by truncating the model output at the point of playback interruption using the approximate ASR transcript timestamp -> token idx mapping described above.

Native VAD models have no notion of barge-in. Since generation is always done in lock-step with client audio, there would only ever be at most `audio_buffer_ms` worth of model output to truncate, which is generally negligible.

#### Tool calling
`/v1/live/sessions` tool calling exists behind `delegation.responses`. On this server, the model is its own Responses backend: `instructions`, `tools`, etc. are rendered by the processor into the model's prompt and sampling parameters, and tool calling is able to be handled via WebSocket events. E.g.:
1. Server send: `session.delegation.created` with `delegation.target: "responses"`
2. Server send: `response.event` carrying that `delegation_id`: `response.created`, etc.
3. Client sends `response.item.create` with the `function_call_output`
4. Client sends `response.create`

A model that does not support tools lists `delegation.responses` in `unsupported_features.session_configs`. `delegation: null` (Live's default when omitted) is the only no-tools configuration; an explicit `delegation: {type: "client"}` is unsupported (see Non-Goals).

#### `audio_buffer_ms`
For `vad: native` only. Buffers this many milliseconds of user audio chunks in the API server before resuming generation.
- MiniCPM-o: 1000 ms
- PersonaPlex: 80 ms

#### `logits_processor`
A map from stage id to a class inheriting vLLM `LogitsProcessor`, installed into that stage's vLLM engine through the existing per-stage `logits_processors` engine arg. Used to enforce MiniCPM-o's duplex chat format.

##### MiniCPM-o LogitsProcessor
In conjunction with the MiniCPM-o LiveSessionProcessor (which simply formats the system prompt and prepends `<unit>` or `<unit/><unit>` upon receiving user audio), enforces the MiniCPM-o duplex chat format via logit masking. See `<file=notes/minicpmo-duplex-logits-processor.py>`.

#### LiveSessionProcessor
The `live_session_processor` provided in the pipeline config is a server-side processor that implements the model-specific operations for formatting new client data and context edits (e.g. interruption truncation, left-trimming when model length is exhausted, etc.).

A generic processor is provided for external VAD models that can rely entirely on their provided ChatML template for rendering (e.g. Qwen3-Omni).

#### PersonaPlex `multi_modal_data_bypass`
To support PersonaPlex's parallel streams architecture, the API server will submit a single placeholder token per generation step, providing the actual inputs via `multi_modal_data` to `AsyncOmni.generate()`. In this case, the model is expected to return a stream per modality (including text) via `multimodal_outputs` which the API server maintains as history per-stream. All are resubmitted after mutation via `multi_modal_data` again with `len(history)` placeholder tokens, analogous to a regular request resume-after-edit.

PersonaPlex's pipeline will be modified to have a `"live_session_mode"`, where it reads its input directly from `multi_modal_data` instead of the conventional input paths. It maintains per-stream context internally with `model_intermediate_buffer` and will reload full edited history when provided in `multi_modal_data`.

#### Unsupported Feature Registration
A `PipelineConfig.live_session_config` registers a list of `/v1/live/sessions` WS client event types and session configurations that are unsupported by this model in `unsupported_features`. Used by MiniCPM-o and PersonaPlex to block tool calling.

When the API server receives a matching event or session config it answers with `error` event.

#### 3.4.1 Dependencies
- API server: openai >= 3.18.0
- SileroVAD service: onnxruntime >= 1.23.2, istupakov/silero-vad-onnx (https://huggingface.co/istupakov/silero-vad-onnx/tree/8b14476858ef240c50b3884bb38cc67290c1cc70)
- Whisper service: vllm, an `openai/whisper-*` checkpoint

#### 3.4.2 Risks
- External services are a new design pattern and may confuse vLLM-Omni server administrators
- /v1/live/sessions is in alpha and may change
- Generic external VAD `LiveSessionProcessor` may break when adding new models with untested ChatML content
- Potentially high prefill overhead on barge-in resubmit path
- The resumable request resubmission path is untested for multi-stage models

### 3.5 Performance

- **Existing fullduplex model performance**: Real-time factor and TTFA within 5% of OmniDuplexSessionHandler on MiniCPM-o and Qwen3-Omni
- **New fullduplex model performance**: RTF <= 1.0 for Nemotron and PersonaPlex

## 4. Correctness & Testing plans

Runtime validation:
- Event format correctness is ensured with Pydantic models

End-to-end tests:
- Per-model tests play a client event sequence against a running vLLM-Omni server and ensure well-formed output

Unit tests, one per component of the stack:
- Audio codec: PCM16 at 16/24 kHz, `pcmu`, `pcma` round trips and resampling against reference arrays
- Session config: `session.start` defaults and echo, the `unsupported_features` rejection matrix, `delegation` handling (`null` accepted, `client` rejected, `responses` gated per model)
- VAD client: against a fake VAD service, synthetic speech and silence fixtures produce speech start/end; mute resets state
- ASR client: against a fake ASR service, transcript timestamp -> token index mapping tolerates ASR and tokenizer drift; input transcript deltas per turn
- Output pacer: with a fake clock, deltas release at 1x; interrupt discards unreleased audio and reports the exact released cursor
- Generic `LiveSessionProcessor`: initial render from `instructions` plus `input`; generation-turn chunk re-emits the dropped stop token; left-trim keeps tool-call pairs and the system prompt; interruption truncation at a cursor; `instructions.append` re-render
- MiniCPM-o processor and logits processor: unit prefix formatting; FSM mask per state; batch add/remove/move
- PersonaPlex bypass: placeholder-plus-audio chunk construction; `model_intermediate_buffer` seeding on resume; dual-history resubmit after mutation
- Resumable request driver: with a fake `AsyncOmni`, append order, restart (abort, new request id, full re-render), end sentinel on close, sampling-parameter constraints enforced
- Tool calling: thinker tool-call markup becomes `session.delegation.created` plus nested Responses events; `function_call_output` and `response.create` continuation; unknown `call_id` errors
- Session lifecycle: `session_lifetime_s` expiry emits `session.closed` with `expired`; close cleanup; usage accounting
- Event serialization: every emitted event validates against `openai.types.live`

## 5. Open Questions & Discussions
- Will there eventually be support for helper models in vLLM-Omni that we should plan to eventually migrate the external services to?
- Are there any metrics not relative to the OmniDuplexSessionHandler stack that this implementation should be evaluated against?

## 6. References

- OpenAI OpenAPI spec (Live API on `main` at commit `7de0436`, 2026-09-22): https://github.com/openai/openai-openapi
- `openai`: https://pypi.org/project/openai/3.18.0/
- SileroVAD: https://github.com/snakers4/silero-vad
- vLLM-Omni fullduplex design doc: `docs/design/fullduplex.md`

### Feedback Period.

2 weeks

### CC List.

@Sy0307 @NickCao @alex-jw-brooks @amy-why-3459 @natureofnature 

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://vllm-omni.readthedocs.io), which can answer lots of frequently asked questions.
